### PR TITLE
Fix brake function, add softStart (accelerate from zero)

### PR DESCRIPTION
**Pull request by ronso0**
Continuing to fix https://bugs.launchpad.net/mixxx/+bug/1571442
"Brake effect initial speed and factor"Initial speed was fixed in #1072.
This PR makes engine.brake() actually take the deceleration factor into account by initializing AlphaBetaFilter with custom beta (How was that done originally at the time the docuwiki was written?).
Additionally, one can now call script.brake(..., factor) in controller script.

### CODE REVIEW

File: res/controllers/common-controller-scripts.js
Comment (Be-ing) [inline res/controllers/common-controller-scripts.js]: This will not work as you expect because of JavaScript's type coercion. Use `factor === undefined`

**Comment by ronso0** [comment]
Should I remove the whole m_rampFactor from brake() function?
I don't see how it affects the brake effect at all.

**Comment by ronso0** [comment]
..forget about that.
It can be used when ramping to 1 like with a softStart function to accelerate to 1, see
https://bugs.launchpad.net/mixxx/+bug/1643720
Although I would probably create a whole new function instead of making brake() more complex.

**Comment by ronso0** [comment]
Please don't merge yet, I'll extend this PR with further clean-up and softStart() function.

**Comment by ronso0** [comment]
...and updated common-controller-scripts.js accordingly

**Comment by ronso0** [comment]
*bump*
I'm done, this can be merged if it's okay. Travis errors aren't related to this code, as far as I can see.
Is clear how softStart() works or should I add comments?

File: res/controllers/common-controller-scripts.js
Comment (Be-ing) [inline res/controllers/common-controller-scripts.js]: use `===`

File: res/controllers/common-controller-scripts.js
Comment (Be-ing) [inline res/controllers/common-controller-scripts.js]: use `!==`

File: res/controllers/common-controller-scripts.js
Comment (Be-ing) [inline res/controllers/common-controller-scripts.js]: use `!==`

**Comment by Be-ing** [comment]
Noticed some small JavaScript nitpicks glancing at the code. I'll give this a try soon.

**Comment by ronso0** [comment]
Okay, thank you! updated
Those === and !== are actually copy/paste from existing code.

softStart() is in general like brake() but has different rampTo[] value and turns on 'play'.
scratchProcess() now considers m_softStartActive[] and uses rampTo[] value as target rate when to turn off scratching.

**Review by Be-ing — CHANGES_REQUESTED** [review]
I tested this and compared it to current master. I confirm that it fixes [the bug](https://bugs.launchpad.net/mixxx/+bug/1571442).

It would be nice if the deck was stopped when there is practically no audio coming through. I tried using script.brake() with really low factors (< .3). I held down the button for a long time after sound stopped coming through, but the deck did not actually stop; when I let go of the button, it started playing at its normal speed again. If you don't want to solve this (or don't want to do it right now), that's fine. I think this PR could be merged without that.

File: res/controllers/common-controller-scripts.js
Comment (ronso0) [inline res/controllers/common-controller-scripts.js]: Indeed. Fixed.

**Comment by ronso0** [comment]
> It would be nice if the deck was stopped when there is practically no audio coming through. I tried using script.brake() with really low factors (< .3)

Yeah, that would be nice! So, there had to be a minimum rate, something like setting rampTo to 0.001 for brake(). I'll experiment with what sounds reasonable.
Now that you mention it, I guess the same should be applied to softStart() when it's triggered with really low factors. I'm confident I can finish this until the weekend, then it can be merged.

> when I let go of the button, it started playing at its normal speed again.

I hope you don't mind that, it was intended for original brake/spinback. I'm quite sure you noticed that you can map start and stop of brake() or softStart() to different buttons, and then sending just NoteOn (NoteOff respetively).
I can add that explicit info to the wiki to clarify the usage of those two functions. That would also solve this bug I guess https://bugs.launchpad.net/mixxx/+bug/1090780

**Comment by Be-ing** [comment]
> I hope you don't mind that, it was intended for original brake/spinback. I'm quite sure you noticed that you can map start and stop of brake() or softStart() to different buttons, and then sending just NoteOn (NoteOff respetively).

Yeah, they could be mapped that way. They could also be mapped to replace a normal play button.

**Comment by Be-ing** [comment]
@Pegasus-RPG any thoughts on this?

**Comment by ronso0** [comment]
>>It would be nice if the deck was stopped when there is practically no audio coming through. I tried using script.brake() with really low factors (< .3)

>Yeah, that would be nice! So, there had to be a minimum rate, something like setting rampTo to 0.001 for brake(). I'll experiment with what sounds reasonable.

I tried to set the minimum rate via rampTo but that would make the unwanted effect even worse.
I'll check if setting a minimum rate in scratchProcess (for brake/spinback only)
Will be back in a few days, I guess.

**Comment by Be-ing** [comment]
I think this should be merged now. Improvements for handling very low factors can come in a future PR.

**Comment by ronso0** [comment]
Yup! As soon as this is merged, I'll complete the wiki, including a hint for the low factors.

File: res/controllers/common-controller-scripts.js
Comment (radusuciu) [inline res/controllers/common-controller-scripts.js]: Indent issue?

Comment (ronso0) [inline res/controllers/common-controller-scripts.js]: Indeed, those 3 lines above had tabs. Fixed.

**Comment by daschuer** [comment]
Thank your for the PR and the review!

**Comment by Pegasus-RPG** [comment]
Especially since features like this are nuanced and carefully determined, PLEASE add tests to ensure that future changes don't adversely affect the behavior.

**Comment by ronso0** [comment]
Cool, thanks for merging! I'll add info to the wiki.
@Pegasus-RPG Unfortunately I have no idea how to write tests. Though, I'm eager to learn, so it would be great if someone else writes those test and I follow that.
There's no test for brake function so far. Because that it's not necessary/possible or just that nobody wrote one yet?

**Comment by Be-ing** [comment]
There's no test because no one wrote one yet. There is a [quick intro](http://mixxx.org/wiki/doku.php/unit_tests) to writing tests on the wiki.

**Pull request merged**
